reco_utils/recommender/newsrec/IO/news_iterator.py

```diff
@@ -0,0 +1,308 @@
+            click_news_index
+        )
+
+
+    def load_data_from_file(self, infile):
+        """Read and parse data from a file.
+        
+        Args:
+            infile (str): text input file. Each line in this file is an instance.
+
+        Returns:
+            obj: An iterator that will yields parsed results, in the format of graph feed_dict.
+        """
+        label_list = []
+        imp_indexes = []
+        user_indexes = []
+        candidate_news_indexes = []
+        click_news_indexes = []
+        cnt = 0
+
+        with tf.gfile.GFile(infile, "r") as rd:
+            while True:
+                line = rd.readline()
+                if not line:
+                    break
+
+                label, imp_index, user_index, candidate_news_index, click_news_index = self.parser_one_line(line)
+
+                candidate_news_indexes.append(candidate_news_index)
+                click_news_indexes.append(click_news_index)
+                imp_indexes.append(imp_index)
+                user_indexes.append(user_index)
+                label_list.append(label)
+
+                cnt += 1
+                if cnt >= self.batch_size:
+                    yield self._convert_data(
+                        label_list, 
+                        imp_indexes,
+                        user_indexes,
+                        candidate_news_indexes, 
+                        click_news_indexes)
+                    candidate_news_indexes = []
+                    click_news_indexes = []
+                    label_list = []
+                    imp_indexes = []
+                    user_indexes = []
+                    cnt = 0
+
+    def _convert_data(
+        self,
+        label_list,
+        imp_indexes,
+        user_indexes,
+        candidate_news_indexes,
+        click_news_indexes
+    ):
+        """Convert data into numpy arrays that are good for further model operation.
+        
+        Args:
+            label_list (list): a list of ground-truth labels.
+            imp_indexes (list): a list of impression indexes.
+            user_indexes (list): a list of user indexes.
+            candidate_news_indexes (list): the candidate news article's words indices
+            click_news_indexes (list): words indices for user's clicked news articles
+            
+        Returns:
+            dict: A dictionary, contains multiple numpy arrays that are convenient for further operation.
+        """
+    
+        labels = np.asarray(label_list, dtype=np.float32)
+        imp_indexes = np.asarray(imp_indexes, dtype=np.int32)
+        user_indexes = np.asarray(user_indexes, dtype=np.int32)
+        candidate_news_index_batch = np.asarray(
+            candidate_news_indexes, dtype=np.int64
+        )
+        click_news_index_batch = np.asarray(click_news_indexes, dtype=np.int64)
+        return [imp_indexes, user_indexes, click_news_index_batch, candidate_news_index_batch], labels
+
+
+class NewsTestIterator(BaseIterator):
```

Comment:
is this the same as NewsTrainIterator with npratio=0? Do we need separate classes? Or can we extend that class to reduce redundant code?

Reply:
Thanks for the suggestions. I will try to change the code and combine these two iterators.